File: apps/admin/src/feature-flags.ts
import type { Endpoint, PayloadRequest } from 'payload'
import { AdminHttpError, adminErrorResponse } from './http/errors'

export type CollectionsFeatureFlag =
  | 'cms_auth'
  | 'catalog_scan'
  | 'collections_admin'
  | 'collector_association_read'
  | 'collector_draft_mutation'
  | 'consumer_credentials'
  | 'collections_distribution'

const FLAG_ENVS: Record<CollectionsFeatureFlag, string> = {
  cms_auth: 'CMS_AUTH_ENABLED',
  catalog_scan: 'CATALOG_SCAN_ENABLED',
  collections_admin: 'COLLECTIONS_ADMIN_ENABLED',
  collector_association_read: 'COLLECTOR_ASSOCIATION_READ_ENABLED',
  collector_draft_mutation: 'COLLECTOR_DRAFT_MUTATION_ENABLED',
  consumer_credentials: 'CONSUMER_CREDENTIALS_ENABLED',
  collections_distribution: 'COLLECTIONS_DISTRIBUTION_ENABLED',
}

function productionLike(): boolean {
  const nodeEnv = process.env.NODE_ENV?.trim().toLowerCase()
  const deployment = process.env.ENVIRONMENT?.trim().toLowerCase()

  if (nodeEnv === 'production' || deployment === 'production' || deployment === 'staging') return true
  if (nodeEnv || deployment) return false
  // Match FastAPI's fail-safe environment policy: an unclassified deployment
  // must not silently enable staged production capabilities.
  return true
}

export function featureEnabled(name: CollectionsFeatureFlag): boolean {
  const envName = FLAG_ENVS[name]
  const raw = process.env[envName]?.trim().toLowerCase()
  if (!raw) return !productionLike()
  if (raw === 'true') return true
  if (raw === 'false') return false
  throw new Error(`${envName} must be true or false`)
}

export function requireFeature(name: CollectionsFeatureFlag): void {
  if (!featureEnabled(name)) throw new AdminHttpError(503, 'feature_disabled', { flag: name })
}

/** Wrap Payload custom endpoints without trusting client/UI feature state. */
export function guardFeatureEndpoints(name: CollectionsFeatureFlag, endpoints: Endpoint[]): Endpoint[] {
  return endpoints.map((endpoint) => {
    const handler = endpoint.handler
    return {
      ...endpoint,
      handler: async (request: PayloadRequest) => {
        try {
          requireFeature(name)
        } catch (error) {
          return adminErrorResponse(error)
        }
        return handler(request)
      },
    }
  })
}

File: concierge-api-v3/app/core/feature_flags.py
"""Server-side feature flags for the staged Collections rollout.

Staging/production fail closed when an override is absent. Development and test
stay enabled by default so existing local workflows are not silently disabled.
The versioned ownership/removal metadata lives in config/collections-feature-flags.json.
"""

from __future__ import annotations

import os
from collections.abc import Callable

from fastapi import HTTPException, status

_FLAG_ENVS = {
    "cms_auth": "CMS_AUTH_ENABLED",
    "catalog_scan": "CATALOG_SCAN_ENABLED",
    "collections_admin": "COLLECTIONS_ADMIN_ENABLED",
    "collector_association_read": "COLLECTOR_ASSOCIATION_READ_ENABLED",
    "collector_draft_mutation": "COLLECTOR_DRAFT_MUTATION_ENABLED",
    "consumer_credentials": "CONSUMER_CREDENTIALS_ENABLED",
    "collections_distribution": "COLLECTIONS_DISTRIBUTION_ENABLED",
}


def _environment() -> str:
    return os.getenv("ENVIRONMENT", "development").strip().lower()


def enabled(name: str) -> bool:
    try:
        env_name = _FLAG_ENVS[name]
    except KeyError as exc:
        raise RuntimeError(f"Unknown Collections feature flag: {name}") from exc

    raw = os.getenv(env_name)
    if raw is None or not raw.strip():
        return _environment() not in {"staging", "production"}

    normalized = raw.strip().lower()
    if normalized == "true":
        return True
    if normalized == "false":
        return False
    raise RuntimeError(f"{env_name} must be true or false")


def require_collection_flag(name: str) -> None:
    if not enabled(name):
        raise HTTPException(
            status_code=status.HTTP_503_SERVICE_UNAVAILABLE,
            detail={"code": "feature_disabled", "flag": name},
        )


def collection_flag_dependency(name: str) -> Callable[[], None]:
    def dependency() -> None:
        require_collection_flag(name)

    return dependency
